### ISSUE TITLE
AWS PowerShell NetCore module does not respect $env:AWS_PROFILE

### ISSUE THREAD

**Issue by bgshacklett** [aws/aws-tools-for-powershell#8]
## Expected Behavior
Cmdlets in the AWSPowerShell.NetCore module should use the profile specified in $env:AWS_PROFILE.

## Current Behavior
Cmdlets in the AWSPowerShell.NetCore module do _not_ respect the $env:AWS_PROFILE variable.

## Possible Solution
When the variable is specified, its value should be used as the value of "ProfileName"

## Steps to Reproduce (for bugs)
1. Create a named profile in ~/.aws/credentials
2. Run a Cmdlet which allows specifying a profile via the `-ProfileName` parameter, but do not specify a value for that parameter. 

## Context
* Writing documentation which targets both PowerShell and bash environments is more cumbersome because credentials are specified differently.
* Explaining the use of the module to peers can be problematic because some environment variables are supported by the Cmdlets and some are not.
* Tooling which indicates what profile is currently in use is significantly harder to write; the value needs to be maintained and managed as separate state rather than assuming that this variable is authoritative.
* To my knowledge, it is not possible to check the currently active profile after running `Set-AWSCredential`

## Your Environment

* AWSPowerShell.NetCore version used: 3.3.365.0
* Service assembly and version used: n/a
* Operating System and version: Windows 10 1703
* Visual Studio version: n/a
* Targeted .NET platform: n/a

**Comment by matteo-prosperi** [comment]
Thanks for suggesting this. We will consider for the next major version.
We cannot change this in a routine update because it has major backward compatibility concerns: a user who opens a PowerShell console today expects to be using his `default` profile, this would change it to what is specified by the `AWS_PROFILE` variable.
Meanwhile you could consider adding `Set-AWSCredential -ProfileName $env:AWS_PROFILE` to your PowerShell profile.

**Comment by tbp105** [comment]
Please update the code or the documentation https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html because the docs claim that this is supported.

**Comment by matteo-prosperi** [comment]
@tbp105 The page https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html refers to using the [AWS CLI](https://aws.amazon.com/cli/) which is a different product that [AWS Tools for PowerShell](https://aws.amazon.com/powershell/). I suppose the confusion derives from the fact that the AWS CLI can be used in a PowerShell session.

**Comment by ashishdhingra** [comment]
Hi @bgshacklett,

Good afternoon.

The documentation link [Credential and profile resolution](https://docs.aws.amazon.com/sdk-for-net/latest/developer-guide/creds-assign.html) specifies support of `AWS_PROFILE` environment variable. Please try using the latest .NET SDK and AWS Tools for Powershell version, and confirm if this is resolved for you.

Thanks,
Ashish

**Comment by bgshacklett** [comment]
On first attempt, this does not appear to have fixed the issue for me. I'll do some further digging as soon as I have a moment.

**Comment by ashishdhingra** [comment]
Hi @bgshacklett,

Good afternoon.

Please confirm if you were able to verify the support for `AWS_PROFILE` environment variable in the latest .NET SDK and AWS Tools for Powershell versions.

Thanks,
Ashish

**Comment by bgshacklett** [comment]
I just tested with version 4.1.8.0 and the behavior does not appear to have changed. I've set `$env:AWS_PROFILE` to a valid profile name, as tested by passing the profile flag in a previous command, but I still receive an error:
```
~> Get-STSCallerIdentity -region us-east-1
Get-STSCallerIdentity: No credentials specified or obtained from persisted/shell defaults.
```

**Comment by ashishdhingra** [comment]
> I just tested with version 4.1.8.0 and the behavior does not appear to have changed. I've set `$env:AWS_PROFILE` to a valid profile name, as tested by passing the profile flag in a previous command, but I still receive an error:
> 
> ```
> ~> Get-STSCallerIdentity -region us-east-1
> Get-STSCallerIdentity: No credentials specified or obtained from persisted/shell defaults.
> ```
> 
> Edit:
> Running the equivalent command via the AWS CLI, immediately afterward, functioned as expected.

@bgshacklett Thanks for your reply. AWS PowerShell tools use .NET Core SDK and it should work as per code https://github.com/aws/aws-sdk-net/blob/a7c91a25e5b78376d1bf91c13f55383518b4a57b/sdk/src/Core/Amazon.Runtime/Credentials/FallbackCredentialsFactory.cs#L56

However, I was able to verify the environment variable `AWS_PROFILE` somehow does not work with latest version of Powershell tools 4.1.8.0. This needs some investigation.

**Comment by ppittle** [comment]
[AWS Tools for PowerShell documentation](https://docs.aws.amazon.com/powershell/latest/userguide/specifying-your-aws-credentials.html) still lists only two supported ways of retrieving credentials, an doesn't list support for `$env:AWS_PROFILE`:

>The AWS Tools for PowerShell can use either of two credentials stores:
>
>- The AWS SDK store, which encrypts your credentials and stores them in your home folder. In Windows, this store is located at: >C:\Users\username\AppData\Local\AWSToolkit\RegisteredAccounts.json.  he AWS SDK for .NET and Toolkit for Visual Studio can also use the AWS SDK store.
>
>- The shared credentials file, which is also located in your home folder, but stores credentials as plain text.  By default, the credentials file is stored here:
>   - On Windows: `C:\Users\username\.aws\credentials`
>   - On Mac/Linux: `~/.aws/credentials`

**Comment by vgribok** [comment]
This issue appears to be **Windows-specific**. On MacOS the `AWS_PROFILE` environment variable does affect the behavior of the AWS PS commands as expected, while on Windows it does not. 

Also, with the AWS IAM Identity Center supporting SSO, along with the CLI `aws sso login` command, the `credentials` file and credential profiles become non-applicable, so the `Set-AWSCredential -ProfileName $env:AWS_PROFILE` is no longer a work-around. 

5 years is a good run for a bug to go unfixed. Let's make AWS PS usable with sso.

**Comment by ashishdhingra** [comment]
Reproducible

**Comment by Betoz** [comment]
Using `AWS PS 4.1.430` in both Windows `PSVersion 7.2.17` and MacOS `PSVersion 7.1.0`.

AWS PS does respect **non-**`default` `$env:AWS_PROFILE` to **SOME** extent.
In my case at least the `endpoint_url` configuration setting gives effect for **non-**`default` `$env:AWS_PROFILE`.

TL;DR; Ctrl+F `VERBOSE: Invoking`.

Also, apparently one can use credentials from one profile (`-ProfileName`) while using `endpoint_url` from another profile `$env:AWS_PROFILE` 🙃

I discovered this while playing with `LocalStack`.

Having the following `config` and `credentials` files
```editor-config
# ~/.aws/config
[default]

[profile localstack]
region = us-east-1
output = json
endpoint_url = http://localhost:4566

[profile someone]
endpoint_url = http://localhost:1337
```
```editor-config
# ~/.aws/credentials
[default]

[localstack]
aws_access_key_id = test
aws_secret_access_key = test

[another]
aws_access_key_id = watwat
aws_secret_access_key = inthe
```
we get
```PowerShell
PS> $DebugPreference = "continue"

PS> Get-SQSQueue -Verbose
Get-SQSQueue: No credentials specified or obtained from persisted/shell defaults.

PS> Set-AWSCredential -ProfileName localstack

PS> Get-SQSQueue -Verbose
DEBUG: Credentials obtained from stored profile named 'localstack'
Get-SQSQueue: No region specified or obtained from persisted/shell defaults.

PS> Set-DefaultAWSRegion -Region us-east-1

PS> Get-SQSQueue -Verbose
DEBUG: Credentials obtained from stored profile named 'localstack'
DEBUG: Region obtained from shell variable $StoredAWSRegion with value 'us-east-1'
VERBOSE: Invoking Amazon Simple Queue Service (SQS) operation 'ListQueues' in region 'us-east-1'
Get-SQSQueue: The security token included in the request is invalid.

PS> $env:AWS_PROFILE = "localstack"

PS> Get-SQSQueue -Verbose
DEBUG: Credentials obtained from stored profile named 'localstack'
DEBUG: Region obtained from shell variable $StoredAWSRegion with value 'us-east-1'
VERBOSE: Invoking Amazon Simple Queue Service (SQS) operation 'ListQueues' on endpoint 'http://localhost:4566/'
Get-SQSQueue: No connection could be made because the target machine actively refused it. (localhost:4566)
```
All above output is identical for both Windows and MacOS, except for the last line, which in MacOS is
```PowerShell
Get-SQSQueue: Connection refused (localhost:4566)
```
(Not important: the last command goes successful if you have `LocalStack` set up.)

Ofc, I wish AWS PS would **ALLOW** usage of credentials and region from non-`default` `$env:AWS_PROFILE`.
```PowerShell
PS> $DebugPreference = "continue"

PS> Get-SQSQueue -Verbose
Get-SQSQueue: No credentials specified or obtained from persisted/shell defaults.

PS> $env:AWS_PROFILE = "localstack"

PS> Get-SQSQueue -Verbose
Get-SQSQueue: No credentials specified or obtained from persisted/shell defaults.
```

## Relevant AWS PS code

Error message `No credentials specified or obtained from persisted/shell defaults.` comes from [Common/BaseCmdlets.cs](https://github.com/aws/aws-tools-for-powershell/blob/c032de16bafaaedb6c5a1286bd90dd857fe49651/modules/AWSPowerShell/Common/BaseCmdlets.cs#L807).
There you also find the `No region specified` error and both `Credentials obtained from` `Region obtained from` DEBUG loggings.

`_CurrentCredentials` and `_RegionEndpoint` get set here. Code fetches these from strictly `default` profile or somewhere else, but **never** `$env:AWS_PROFILE`. See `TryGetCredentials` and `TryGetRegion` in [Common/CredentialsArguments.cs](https://github.com/aws/aws-tools-for-powershell/blob/c032de16bafaaedb6c5a1286bd90dd857fe49651/modules/AWSPowerShell/Common/CredentialsArguments.cs#L98). Nothing continues without `_CurrentCredentials` being set.

`_CurrentCredentials` and `_RegionEndpoint` are later used to instantiate a `Client` in [Cmdlets/SQS/AmazonSQSClientCmdlet.cs](https://github.com/aws/aws-tools-for-powershell/blob/c032de16bafaaedb6c5a1286bd90dd857fe49651/modules/AWSPowerShell/Cmdlets/SQS/AmazonSQSClientCmdlet.cs#L64). Note how a `config` is also used for the `Client` instantiation. This `config` has properties `RegionEndpoint` and `ServiceURL` initially set to the `region` and `endpoint_url` values for the `$env:AWS_PROFILE` profile in `~/.aws/config` (more on this later). `RegionEndpoint` is overriden by `_RegionEndpoint` though.
The `Client` is later passed to `CallAWSServiceOperation` in [Cmdlets/SQS/Basic/Get-SQSQueue-Cmdlet.cs](https://github.com/aws/aws-tools-for-powershell/blob/c032de16bafaaedb6c5a1286bd90dd857fe49651/modules/AWSPowerShell/Cmdlets/SQS/Basic/Get-SQSQueue-Cmdlet.cs#L180).

The first thing `CallAWSServiceOperation` does is to call `Utils.Common.WriteVerboseEndpointMessage` from [Utils/Utils.cs](https://github.com/aws/aws-tools-for-powershell/blob/c032de16bafaaedb6c5a1286bd90dd857fe49651/modules/AWSPowerShell/Utils/Utils.cs#L139) which is from where both ` in region '` ` on endpoint '` VERBOSE loggings origin from, depending on content of `ServiceURL` and `RegionEndpoint`.
This last sentence is done for over `99%` of all Cmdlets (`14043` of `14060` `*-Cmdlet.cs` files). Looks very standard.

## Confirming `RegionEndpoint` and `ServiceURL` taken from `$env:AWS_PROFILE`.

```PowerShell
PS> Import-Module AWS.Tools.SQS
PS> $myAmazonSQSConfig = [Amazon.SQS.AmazonSQSConfig]::new()
PS> $myAmazonSQSConfig.RegionEndpoint
PS> $myAmazonSQSConfig.ServiceURL
PS> $env:AWS_PROFILE = "localstack"
PS> $myAmazonSQSConfig = [Amazon.SQS.AmazonSQSConfig]::new()
PS> $myAmazonSQSConfig.RegionEndpoint

SystemName         : us-east-1
OriginalSystemName : us-east-1
DisplayName        : US East (N. Virginia)
PartitionName      : aws
PartitionDnsSuffix : amazonaws.com

PS> $myAmazonSQSConfig.ServiceURL
http://localhost:4566/
```

## Mixing `-ProfileName` and `$env:AWS_PROFILE`

```PowerShell
PS> $DebugPreference = "continue"

PS> Get-SQSQueue -Verbose
Get-SQSQueue: No credentials specified or obtained from persisted/shell defaults.

PS> Get-SQSQueue -Verbose -ProfileName "another"
DEBUG: Credentials obtained from stored profile named 'another'
Get-SQSQueue: No region specified or obtained from persisted/shell defaults.

PS> Get-SQSQueue -Verbose -ProfileName "another" -Region eu-central-1
DEBUG: Credentials obtained from stored profile named 'another'
DEBUG: Region obtained from region parameter with value 'eu-central-1'
VERBOSE: Invoking Amazon Simple Queue Service (SQS) operation 'ListQueues' in region 'eu-central-1'
Get-SQSQueue: The security token included in the request is invalid.

PS> $env:AWS_PROFILE = "someone"

PS> Get-SQSQueue -Verbose -ProfileName "another" -Region eu-central-1
DEBUG: Credentials obtained from stored profile named 'another'
DEBUG: Region obtained from region parameter with value 'eu-central-1'
VERBOSE: Invoking Amazon Simple Queue Service (SQS) operation 'ListQueues' on endpoint 'http://localhost:1337/'
Get-SQSQueue: No connection could be made because the target machine actively refused it. (localhost:1337)
```

## Conclusion

AWS PS does not care about credentials and region configuration settings from **non-**`default` `$env:AWS_PROFILE`.
It does however care about other configuration settings from **non-**`default` `$env:AWS_PROFILE`.
You can mix credentials from one profile with configuration settings from another profile.

It would be super great if `$env:AWS_PROFILE` is also respected for credentials and region, and that `-ProfileName` is used for not just credentials but also configuration settings and that it takes precedence over `$env:AWS_PROFILE`.
This just feels a lot more intuitive than current behavior.

**Comment by eddiemcs3** [comment]
We are currently looking to include this in our next major version update.

**Comment by ashishdhingra** [comment]
This has been fixed in **AWS Tools for PowerShell** version `5.0.0-preview003`.

**Instructions:** (taking `AWS.Tools.Route53` as an example)
- Install using below commands
  ```PowerShell
  Install-Module -Name AWS.Tools.Common -RequiredVersion 5.0.0-preview003 -AllowPrerelease
  Install-Module -Name AWS.Tools.Route53 -RequiredVersion 5.0.0-preview003 -AllowPrerelease
  ```
- You would need to manually import `AWS.Tools.Common` (refer [V5 Development Tracker](https://github.com/aws/aws-tools-for-powershell/issues/357))
  ```PowerShell
  Import-Module AWS.Tools.Common -RequiredVersion '5.0.0'
  Import-Module AWS.Tools.Route53 -RequiredVersion '5.0.0'
  ```
- Set a named profile (e.g. `default-renamed`) in `~/.aws/credentials` file with the configured credentials. Make sure that there is no `default` profile configured.
- Executing the command `(Get-R53ResourceRecordSet -HostedZoneId Z05229801H2FUQ0KFG7R8).ResourceRecordSets` gives below error:
  ```PowerShell
  Get-R53ResourceRecordSet: Unable to get IAM security credentials from EC2 Instance Metadata Service.
  ``` 
- Now in current PowerShell session, set the environment variable `AWS_PROFILE` using command ` $env:AWS_PROFILE="default-renamed"`.
- Executing `(Get-R53ResourceRecordSet -HostedZoneId <<Hosted-Zone-Id>>).ResourceRecordSets` (specify value for placeholder `<<Hosted-Zone-Id>>`) returns below output:
  ```PowerShell
  Name                    : testdomain.com.
  Type                    : NS
  SetIdentifier           :
  Weight                  :
  Region                  :
  GeoLocation             :
  Failover                :
  MultiValueAnswer        :
  TTL                     : 172800
  ResourceRecords         : {Amazon.Route53.Model.ResourceRecord, Amazon.Route53.Model.ResourceRecord, 
  Amazon.Route53.Model.ResourceRecord,
                          Amazon.Route53.Model.ResourceRecord}
  AliasTarget             :
  HealthCheckId           :
  TrafficPolicyInstanceId :
  CidrRoutingConfig       :
  GeoProximityLocation    :
  
  Name                    : testdomain.com.
  Type                    : SOA
  SetIdentifier           :
  Weight                  :
  Region                  :
  GeoLocation             :
  Failover                :
  MultiValueAnswer        :
  TTL                     : 900
  ResourceRecords         : {Amazon.Route53.Model.ResourceRecord}
  AliasTarget             :
  HealthCheckId           :
  TrafficPolicyInstanceId :
  CidrRoutingConfig       :
  GeoProximityLocation    :
  ```

Closing this issue. Feel free to comment in case you notice that the issue has not been fixed.
